apps/web/src/app/[locale]/[username]/_components/dashboard/in-progress-tab.tsx

```diff
@@ -1,39 +1,55 @@
 import {
   Table,
   TableBody,
-  TableCaption,
   TableCell,
   TableHead,
   TableHeader,
   TableRow,
 } from '@repo/ui/components/table';
+import { prisma } from '@repo/db';
+import { cache } from 'react';
+import Link from 'next/link';
+import { getRelativeTime } from '~/utils/relativeTime';
 
-const challengesList = [
-  {
-    challengeTitle: 'Extreme Challenge 1',
-    challengeStatus: 'In-Progress',
-  },
-  {
-    challengeTitle: 'Extreme Challenge 4',
-    challengeStatus: 'In-Progress',
-  },
-];
+const getInProgressSubmissions = cache(async (id: string) => {
+  return await prisma.submission.findMany({
+    where: {
+      userId: id,
+      isSuccessful: false,
+    },
+    orderBy: [
+      {
+        createdAt: 'desc',
+      },
+    ],
+    take: 25,
+    include: {
+      challenge: true,
+    },
+    distinct: 'challengeId',
+  });
```

Comment:
I don't think this logic is correct. By definition, "in-progress" means something that has been started but not completed yet. Your query is just returning a list of submissions where the user has has made an attempt and failed previously, but doesn't take into account the fact whether the user might have actually also solved it in a different attempt.

For example if the submissions table looked like this (simplified):
```typescript
[
  { id: 1, challengeId: 1, userId: 1, isSuccesful: false}, 
  { id: 2, challengeId: 1, userId: 1, isSuccesful: true}
]
```

As you can see, user 1 attempted challenge 1 and failed. Challenge 1 is now in-progress. User 1 made another attempt and submitted a correct solution and the submission is successful. Challenge 1 is now completed.

For this example, your query will return challenge 1 still because it has `isSuccessful: false`.

There are multiple ways to solve this efficiently with Prisma, but the general question is: Find all challenges that have at least one or more submissions AND none of them are successful.

I might be wrong with the syntax, but I think this query (or something similar) should work:
```typescript
return await prisma.challenge.findMany({
  where: {
    submissions: {
      none: {
        where: {
          userId: id,
          isSuccessful: true,
        },
      },
    },
  },
});
```

Read more about Prisma relation filters [here](https://www.prisma.io/docs/concepts/components/prisma-client/relation-queries#relation-filters)

Reply:
Again, I might be wrong with Prisma syntax (a bit rusty), but I think a nested relationship query can be something like this?

```typescript
return await prisma.challenge.findMany({
  where: {
    submissions: {
      none: {
        where: {
          userId: id,
          isSuccessful: true,
        },
      },
    },
  },
  select: {
    id: true,
    submissions: {
      orderBy: {
        createdAt: 'desc'
      },
      take: 1,
      select: {
        createdAt: true
      }
    }
  }
});
```

This query will return a list of challenges that are in-progress for the current user, and also return the createdAt field of the latest failed submission.